src/mesh.f90

```diff
@@ -0,0 +1,375 @@
+    integer, dimension(3) :: pnext ! rank ID of the previous rank in each direction
+    integer, dimension(3) :: pprev ! rank ID of the next rank in each direction
+  contains
+    procedure :: is_root ! returns if the current rank is the root rank
+  end type
+
+
+  ! The mesh class stores all the information about the global and local (due to domain decomposition) mesh
+  ! It also includes getter functions to access some of its parameters
+  type :: mesh_t
+    integer, dimension(3), private :: dims_global ! global number of vertices in each direction
+
+    integer, dimension(3), private :: vert_dims_padded ! local domain size including padding (cartesian structure)
+    integer, dimension(3), private :: vert_dims ! local number of vertices in each direction without padding (cartesian structure)
+    integer, dimension(3), private :: cell_dims ! local number of cells in each direction without padding (cartesian structure)
+    logical, dimension(3), private :: periodic_dir ! Whether or not a direction has a periodic BC
+    integer, private :: sz
+    class(geo_t), allocatable :: geo ! object containing geometry information
+    class(parallel_t), allocatable :: par ! object containing parallel domain decomposition information
+  contains
+    procedure :: get_SZ
+
+    procedure :: get_n_groups_dir
+    procedure :: get_n_groups_phi
+    generic :: get_n_groups => get_n_groups_dir, get_n_groups_phi
+
+    procedure :: get_dims_dir
+    procedure :: get_dims_phi
+    procedure :: get_dims_phi_dataloc
+    generic :: get_dims => get_dims_dir, get_dims_phi, get_dims_phi_dataloc
+
+    procedure :: get_n_dir 
+    procedure :: get_n_phi
+    generic :: get_n => get_n_dir, get_n_phi
+
+    procedure :: get_padded_dims_phi
+    procedure :: get_padded_dims_dir
+    generic :: get_padded_dims => get_padded_dims_dir, get_padded_dims_phi
+  
+    procedure :: get_coordinates
+
+    procedure :: set_sz
+    procedure :: set_padded_dims
+  end type mesh_t
+
+  interface mesh_t
+    module procedure mesh_init
+  end interface mesh_t
+
+  contains 
+
+  function mesh_init(dims_global, nproc_dir, L_global) result(mesh)
+    !! Completely initialise the mesh object. 
+    !! Upon initialisation the mesh object can be read-only and shouldn't be edited
+    !! Takes as argument global information about the mesh like its length, number of cells and decomposition in each direction
+    integer, dimension(3), intent(in) :: dims_global
+    integer, dimension(3), intent(in) :: nproc_dir ! Number of proc in each direction
+    real(dp), dimension(3), intent(in) :: L_global
+    type(mesh_t) :: mesh
+
+    integer :: nx, ny, nz, dir
+    integer :: ierr
+    logical :: is_last_domain
+
+    allocate(mesh%geo)
+    allocate(mesh%par)
+    mesh%dims_global(:) = dims_global
+    ! Geometry
+    mesh%geo%L = L_global
+    mesh%geo%d = mesh%geo%L(:) / mesh%dims_global(:)
```

Comment:
`d` actually depends on periodicity but can be fixed later on as well. This is correct for a periodic case. If not periodic then we need to divide to `mesh%dims_global(dir)-1`, its like the global cell count in that direction.